waiter/src/waiter/reporter.clj

```diff
@@ -0,0 +1,151 @@
+
+(defn- make-metrics-filter
+  [filter-regex]
+  (reify MetricFilter (matches [_ name _] (some? (re-matches filter-regex name)))))
+
+(defn- make-codahale-reporter
+  [^ScheduledReporter scheduled-reporter state period-ms]
+  (.start scheduled-reporter period-ms TimeUnit/MILLISECONDS)
+  (reify CodahaleReporter
+    (close! [_] (.close scheduled-reporter) (reset! state :closed))
+    (state [_] @state)))
+
+
+(defn validate-console-reporter-config
+  "Validates ConsoleReporter settings and sets defaults"
+  [config]
+  (s/validate {(s/required-key :period-ms) schema/positive-int
+               (s/optional-key :filter-regex) s/Regex
+               s/Any s/Any} config))
+
+(defn make-console-reporter
+  "Creates a ConsoleReporter for codahale metrics"
+  ([filter-regex] (make-console-reporter filter-regex nil))
+  ([filter-regex ^PrintStream output]
+   (let [state (atom {})
+         console-reporter (cond-> (ConsoleReporter/forRegistry metrics/default-registry)
+                                  output (.outputTo output)
+                                  filter-regex (.filter (make-metrics-filter filter-regex))
+                                  true (-> (.convertRatesTo TimeUnit/SECONDS)
+                                           (.convertDurationsTo TimeUnit/MILLISECONDS)
+                                           (.build)))]
+     [console-reporter state])))
+
+(defn console-reporter
+  "Creates and starts a ConsoleReporter for codahale metrics"
+  [config]
+  (let [{:keys [period-ms filter-regex]} (validate-console-reporter-config config)
+        [console-reporter state] (make-console-reporter filter-regex)]
+    (make-codahale-reporter console-reporter state period-ms)))
+
+
+
+(defn validate-graphite-reporter-config
+  "Validates GraphiteReporter settings and sets defaults"
+  [config]
+  (s/validate {(s/required-key :period-ms) schema/positive-int
+               (s/optional-key :filter-regex) s/Regex
+               (s/required-key :host) s/Str
+               (s/optional-key :prefix) s/Str
+               (s/required-key :port) schema/positive-int
+               (s/required-key :pickled?) s/Bool
+               s/Any s/Any}
+              (merge {:pickled? true} config)))
+
+(defn make-graphite-reporter
+  "Creates a GraphiteReporter for metrics"
+  ([filter-regex prefix host port pickled?]
+   (let [addr (InetSocketAddress. ^String host ^int port)
+         ^GraphiteSender graphite (if pickled?
+                                    (PickledGraphite. addr)
+                                    (Graphite. addr))]
+     (make-graphite-reporter filter-regex prefix graphite)))
+  ([filter-regex prefix ^GraphiteSender graphite]
+   (let [state (atom {})
+         graphite-wrapper (reify GraphiteSender
+                            (connect [_]
+                              (try (.connect graphite)
+                                   (catch Exception e
+                                     (swap! state #(merge % {
```

Comment:
Nitpick:
`(swap! an-atom #(f % x y z))`
is same as:
`(swap! an-atom f x y z)`

I find the second variant easier to read.

Reply:
Also feels like this `swap!` should be inside a helper function:
```
(fn update-state [graphite successful?]
  (swap! state assoc :failed-writes-to-server (.getFailures graphite)
                     :last-connect-failed-time (t/now)
                     :last-report-successful successful?))
```